packages/next/src/server/after/after-context.ts

```diff
@@ -0,0 +1,189 @@
+    }
+  }
+
+  const runCallbacks = (requestStore: RequestStore) => {
+    if (afterCallbacks.length === 0) return
+
+    const runCallbacksImpl = () => {
+      while (afterCallbacks.length) {
+        const afterCallback = afterCallbacks.shift()!
+
+        const onError = (err: unknown) => {
+          // TODO(after): how do we properly report errors here?
+          console.error(
+            'An error occurred in a function passed to after()',
+            err
+          )
+        }
+
+        // try-catch in case the callback throws synchronously or does not return a promise.
+        try {
+          const ret = afterCallback()
+          if (isPromise(ret)) {
+            waitUntil(ret.catch(onError))
+          }
+        } catch (err) {
+          onError(err)
+        }
+      }
+
+      keepAliveLock.release()
+    }
+
+    const readonlyRequestStore: RequestStore =
+      wrapRequestStoreForAfterCallbacks(requestStore)
+
+    return requestAsyncStorage.run(readonlyRequestStore, () =>
+      cacheScope ? cacheScope.run(runCallbacksImpl) : runCallbacksImpl()
+    )
+  }
+
+  return {
+    enabled: true as const,
+    after: afterImpl,
+    run: async <T>(requestStore: RequestStore, callback: () => T) => {
+      try {
+        return await (cacheScope
+          ? cacheScope.run(() => callback())
+          : callback())
+      } finally {
+        // NOTE: it's likely that the callback is doing streaming rendering,
+        // which means that nothing actually happened yet,
+        // and we have to wait until the request closes to do anything.
+        // (this also means that this outer try-finally may not catch much).
+
+        // don't await -- it may never resolve if no callbacks are passed.
+        void onCloseLazy(() => runCallbacks(requestStore)).catch(
+          (err: unknown) => {
+            console.error(err)
+            // as a last resort -- if something fails here, something's probably broken really badly,
+            // so make sure we release the lock -- at least we'll avoid hanging a `waitUntil` forever.
+            keepAliveLock.release()
+          }
+        )
+      }
+    },
+  }
+}
+
+export function createDisabledAfterContext() {
+  return { enabled: false as const }
+}
+
+/** Disable mutations of `requestStore` within `after()` and disallow nested after calls.  */
+function wrapRequestStoreForAfterCallbacks(
+  requestStore: RequestStore
+): RequestStore {
+  return {
+    get headers() {
+      return requestStore.headers
+    },
+    get cookies() {
+      return requestStore.cookies
+    },
+    get draftMode() {
+      return requestStore.draftMode
+    },
+    assetPrefix: requestStore.assetPrefix,
+    reactLoadableManifest: requestStore.reactLoadableManifest,
+    // make cookie writes go nowhere
+    mutableCookies: new ResponseCookies(new Headers()),
+    afterContext: {
+      enabled: true,
+      after: () => {
+        throw new Error('Cannot call after() from within after()')
```

Comment:
after-within-after is a feature we probably want, but i think it'd be better to leave that for a follow-up PR -- this is big enough as it is.